### ISSUE TITLE
[DX] Consolidate all functions that generate permission matrices, and make code consistent across all modules.

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#5031]
I started working on #5030, but it didn't take long before I started being annoyed by the amount of code duplication and inconsistencies in the implementation of permission matrices in different modules (Contact module settings form, "Permissions" vertical tabs in the vocabulary add/edit form and the content type add/edit form etc.)

Ideally, I would like that we introduced a new Form API element to handle these (), but that may be harder to do. In the meantime, let's please fix the existing mess.

**Comment by klonos** [comment]
PR up for review/testing: https://github.com/backdrop/backdrop/pull/3580

I've introduced a single new function that generates the portion of the form that is the permissions/roles matrix, and I'm calling it from within `contact_form_permissions()`, `taxonomy_vocabulary_form_permissions()`, `node_type_form_permissions()`, as well as from `user_admin_permissions()` (the function that generates the permissions in `admin/config/people/permissions`).

What I would like feedback on:
- I called the new function `user_generate_form_permissions()`. Is that a good name?
- I placed that function in `user.admin.inc`. Is that the most appropriate place?
- I need help/suggestions with the function description, `@param`, and `@return` comments in its docblock.
- anything else I may have missed, or that could be improved.

**Comment by jenlampton** [comment]
> I called the new function `user_generate_form_permissions()`. Is that a good name?

How about `backdrop_generate_permissions_form()`?

> I placed that function in user.admin.inc. Is that the most appropriate place?

If we could put it in `user.module` we wouldn't need to call `module_load_include()` to get it.

> The function accepts an optional boolean parameter called $form_is_new, which can be used to specify whether the form is to create a new item (vocabulary or content type for example), or edit an existing item. Does that parameter name make sense?

I prefer simply `$is_new` for the name.

> I need help/suggestions with the function description, @param, and @return comments in its docblock.

I looked over [what you've already done](https://github.com/backdrop/backdrop/pull/3580/files#diff-7238222145839569d46199bfde427f1d35a2a9562a4c0b5d2d34af4e4efad7c6R460) and it looks fantastic!

**Comment by klonos** [comment]
> How about `backdrop_generate_permissions_form()`?

That makes it imply that it generates a form, whereas what it does is to only create a portion of it, which holds permission checkbox elements. Any others have ideas before I make any changes?

> If we could put it in `user.module` or even `system.module` we wouldn't need to call `module_load_include()` to get it.

Yes, that's why I asked basically for feedback on this 😅 ...I'll wait a while so that others chime in, and will move it to the place that has the most consensus 👍🏼 

> I prefer simply `$is_new` for the name.

Sounds good to me 👍🏼 ...any others?

> I looked over [what you've already done](https://github.com/backdrop/backdrop/pull/3580/files#diff-7238222145839569d46199bfde427f1d35a2a9562a4c0b5d2d34af4e4efad7c6R460) and it looks fantastic!

Thanks 😳

**Comment by klonos** [comment]
I'll resume work on this one. Starting by attempting to rebase first...

**Comment by klonos** [comment]
Thanks for the feedback @kiamlaluno 🙏🏼 ...yes, I was just re-inspecting the code (been a couple of years since I last touched it 😅 ), and I was also thinking to use the verb "build" instead of "generate". Your suggestion for `user_build_permission_checkboxes()` matches what I was thinking 👍🏼 ...either that, or the lengthier (but even more accurate) `user_build_form_permission_checkboxes()`.

Anyway, I just rebased, and I needed to do was:
- fix a typo (yay cspell check! 🎉  and thank you once again @BWPanda and @indigoxela for working on all the automations around our CI and sandboxes 🙏🏼 🙏🏼 🙏🏼 )
- fix a failing test that was expecting a redirect to the vocabulary listing after creating a new vocabulary (we now redirect to the term listing page for the newly-created vocabulary).

All tests are green now (last time I touched this PR we were still using Zen CI ...time flies 😅).

**Comment by klonos** [comment]
...also, here's what I was thinking re the `$form_is_new` variable (which @jenlampton suggested to change to just `$is_new`) and its use in the new function: I would like to "decouple" the function this PR introduces from the intended use of that parameter (whether the form that contains the permission checkboxes creates new items or edits existing ones), as the permission matrices may be used in configuration forms that have nothing to do with creating things - just with "pure" settings. I instead focused on the usage of this parameter/variable within the function itself, which is to determine if a specific role should be assigned all permissions or not. So I would like to work on changing the function from this:

`function user_build_form_permission_checkboxes($permissions = NULL, $form_is_new = FALSE, $role_name = NULL) { ... }`

...to this:

`function user_build_form_permission_checkboxes($permissions = NULL, $role_name = NULL, $all_permissions_roles = array()) { ... }`

So basically change that parameter from a mere boolean "flag", to an array of roles that should be assigned all permissions.

I haven't updated the PR with all that yet, but it is what I am planning to work on next.

**Comment by kiamlaluno** [comment]
`user_build_form_permission_checkboxes()` is an even better name.

**Comment by jenlampton** [comment]
> I instead focused on the usage of this parameter/variable within the function itself, which is to determine if a specific role should be assigned all permissions or not. It basically all boils down to this specific conditional:

As discussed in the weekly dev meeting this week, I don't think this is a good idea. Other than the admin role (which we can determine without passing it in as a parameter) it will be very unlikely that we will want to assign *all* permissions in this matrix to any role, so having an array of roles here not be helpful. 

One thing that might be valuable would be the ability to pass in a set of default values. It would need to account for all possible values, so all roles and all permissions, and each module would need the ability to set which ones are suitable for which. 

I had proposed that we add an alter hook insead, rather than trying to guess how various modules might want to use this mystery feature. Once we see all of the actual use-cases in core for how this might be used, we can re-evaluate and add more parameters if necessary. For now, we should stick with `$is_new`.

I would recommend the function names either of the following
* `user_permission_form_checkboxes()` and `user_permission_form_grid_checkboxes()`
* `user_permission_form_grid()` and `user_permission_form_grid_alter()`

My logic (not necessarily the best...)
* I'm not sure the word 'build' adds much value (but I do prefer it to `generate`).
* what we are building is a `permission form` so I prefer that order than `form permission` 
* `checkboxes` is very specific. if we ever wanted to change this to some other future html element, we wouldn't need to change the function name at that time. (We might want to anyway - but we'd at least have a choice).
* `checkboxes` is a long word

**Comment by kiamlaluno** [comment]
I would add a verb in the function name. Without that, the function name does not make explicit what the function does. With four nouns, it is also difficult to understand which noun the noun adjuncts modify. (Is it *user-permission form grid* or *user-permission form-grid*?)

I would avoid *alter* at the end of the function name, since we have alter hooks, and `user_permission_form_grid_alter()` could be understood as function that calls some alter hooks.

I would opt for `user_permission_build_form()`.

**Comment by klonos** [comment]
OK, I've rebased the PR branch and started working on this again.

> I prefer simply `$is_new` for the name.

I've renamed that variable from `$form_is_new` to `$is_new` 👍🏼 

>> I placed that function in user.admin.inc. Is that the most appropriate place?
>
> If we could put it in `user.module` or even `system.module` we wouldn't need to call `module_load_include()` to get it.

Makes perfect sense 👍🏼 ...I've moved the function into `user.module`.

> - `user_permission_form_grid()` and `user_permission_form_grid_alter()`
> ...
> - I'm not sure the word 'build' adds much value (but I do prefer it to `generate`).

> - what we are building is a `permission form` so I prefer that order than `form permission`
> - `checkboxes` is very specific. ...
> - `checkboxes` is a long word

> I would add a verb in the function name. Without that, the function name does not make explicit what the function does. With four nouns, it is also difficult to understand which noun the noun adjuncts modify. (Is it *user-permission form grid* or *user-permission form-grid*?)

I hear both of you. It's difficult to name things 😞 ...pending further feedback, I have chosen `user_build_form_permissions_grid()` for now.

Still tweaking. Planning to implement other feedback as well, like implementing alter hooks to modify the defaults etc., so this is WIP. Tagging as such.

**Comment by klonos** [comment]
...I'm also glad to report that I have removed another hack that we were repeating everywhere we were rendering these permission matrices. All these nasty things are now gone:

`contact.admin.inc`
```php
  // Note that in contact_theme(), the 'contact_form_permissions' theme callback
  // reuses theme_user_admin_permissions(), to make it identical to the user
  // module permissions page.
  $form['#theme'] = array('contact_form_permissions');
```

`node.types.inc`
```php
  // Note that in node_theme(), the 'node_type_form_permissions' theme callback
  // reuses theme_user_admin_permissions(), to make it identical to the user
  // module permissions page.
  $form['#theme'] = array('node_type_form_permissions');
```

`taxonomy.admin.inc`
```php
  // Note that in taxonomy_theme() the 'taxonomy_vocabulary_form_permissions'
  // theme callback reuses theme_user_admin_permissions(), to make it identical
  // to the user module permissions page.
  $form['#theme'] = array('taxonomy_vocabulary_form_permissions');
```

And now all we have is a single `$form['#theme'] = array('user_admin_permissions');` line in the `user_build_form_permissions_grid()` function in `user.module` 😉 

Another DX+ win! 🙂 

Other improvements include:
- inline documentation to explain what's happening in various places where things seemed "magic" before
- naming some variables properly (instead of adding yet more words in the cspell dictionary), which should make certain things self-explanatory now
- simplification of the replacement logic/functions for the placeholder permissions in `node_type_form_submit()` and `taxonomy_form_vocabulary_submit()` (when creating new vocabularies or content types, and the `elseif ($status == SAVED_NEW) {` condition applies)
